src/kernels/common/baseJupyterSession.ts

```diff
@@ -335,14 +343,33 @@ export abstract class BaseJupyterSession implements IBaseKernelConnectionSession
                     }
                 };
                 session.kernel.statusChanged?.connect(handler);
+                disposables.push(
+                    new Disposable(() => swallowExceptions(() => session.kernel?.statusChanged?.disconnect(handler)))
+                );
                 if (session.kernel.status == 'idle') {
                     deferred.resolve(session.kernel.status);
                 }
-                const result = await Promise.race([deferred.promise, sleep(timeout)]);
-                session.kernel.statusChanged?.disconnect(handler);
+                // Check for possibility that kernel has died.
+                const sessionDisposed = createDeferred<unknown>();
+                session.disposed.connect(sessionDisposed.resolve, sessionDisposed);
+                disposables.push(
+                    new Disposable(() =>
+                        swallowExceptions(() => session.disposed.disconnect(sessionDisposed.resolve, sessionDisposed))
+                    )
+                );
+                const sleepPromise = sleep(timeout);
+                sessionDisposed.promise.ignoreErrors();
+                sleepPromise.ignoreErrors();
+                deferred.promise.ignoreErrors();
+                const result = await Promise.race([deferred.promise, sleepPromise, sessionDisposed.promise]);
+                if (session.isDisposed) {
```

Comment:
This is what was missing, at this point its possible the session has been disposed.
We were getting errors because kernel was disposed and `session.kernel` was undefined.

We would be faster in handling cases where restarts fail before the timeout. 
Before this change we'd wait for 60s, now if the session dies before that we'd exit this block much sooner than 60s.